File: add_two_nums/addTwoNums.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        // set a pointer to the front of the linked list,
        // as well as a pointer to the current node
        ListNode first = null, current = null;
        int sum, sumBit, carryBit = 0;
        
        // iterate until we have reached the end of both lists
        while (l1 != null | l2 != null) {
            
            // if we have reached the end of one number
            if (l1 == null)
                sum = l2.val + carryBit;
            else if (l2 == null)
                sum = l1.val + carryBit;
            
            // both numbers still have digits
            else 
                sum = l1.val + l2.val + carryBit;
            
            sumBit = sum % 10;
            carryBit = sum / 10;
            
            // create nodes in the linked list
            if (first == null) {
                first = new ListNode(sumBit);
                current = first;
            }
            else {
                current.next = new ListNode(sumBit);
                current = current.next;
            }
            
            // update pointers 
            if (l1 != null)
                l1 = l1.next;
            if (l2 != null)
                l2 = l2.next;
        }
        
        // if the carry is non-zero, create a final node
        if (carryBit > 0) {
            current.next = new ListNode(carryBit);
            current.next.next = null;
        }
        
        return first;
    }
}

File: add_two_nums/add_two_nums.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        
        # return value and an a current value we use
        # to traverse
        ret = ListNode(0)
        cur = ret
        
        # keep going while we have something to add
        while(l1 or l2):
            
            # sum the current value (0 or 1 if a carry) and l1 and/or l2 
            # if they exist, and set current value to mod 10 of that
            s = cur.val + (l1.val if l1 else 0) + (l2.val if l2 else 0)
            cur.val = s % 10

            # next node starts on 1 if we carry
            if s >= 10:
                cur.next = ListNode(1)
                
            # return if we can't track any further
            if (not (l1 and l1.next) and not (l2 and l2.next)):
                return ret
            
            # otherwise increment as appropriate for l1, l2, cur
            else:
                l1 = l1.next if l1 else None
                l2 = l2.next if l2 else None
                cur.next = cur.next or ListNode(0)
                cur = cur.next